File: solve/lua/factoradic.lua
#! /usr/bin/lua

require "common"

f = {}

function permute(k)
	for i = 1, n - 1 do
		local pos = math.floor(k / f[n - i]) + i
		local tmp = a[pos] 
		for j = pos, i + 1, -1 do 
			a[j] = a[j - 1]
		end 
		a[i] = tmp
		k = k % f[n - i]
	end
end

skeleton(ORDERED, function ()
	f[0] = 1
	for i = 1, n do
		f[i] = f[i - 1] * i
	end

	local top = f[n]
	for k = 0, top - 1 do
		undo()
		permute(k)
		output()
	end
end)

File: solve/cpp/factoradic.cc
#include "common.hh"

int f[MAX];

void precompute()
{
    f[0] = 1;
    for (int i = 1; i != n; ++i)
    {
        f[i] = f[i - 1] * i;
    }
}

void permute(int k)
{
    for (int i = 0; i != n; ++i)
    {
        int j = i + k/f[n - 1 - i];

        char c = a[j];
        for (; j != i; --j)
        {
            a[j] = a[j - 1];
        } 
        a[i] = c;

        k %= f[n - 1 - i];
    }
}

int main(int argc, char *argv[])
{
    if (init(argc, argv, ORDERED) == 0)
    { 
        precompute(); 
        for (int i = 0; i != f[n - 1]*n; ++i)
        {
            undo();
            permute(i);
            output();
        } 
        return final();       
    }
    return 1; 
}
